test/integration_test/node_pdb_test.go

```diff
@@ -0,0 +1,205 @@
+				return true
+			}
+			k8sVersion, _ := version.NewVersion(kbVer)
+			return ci_utils.PxOperatorVersion.LessThan(ci_utils.PxOperatorVer24_2_0) || k8sVersion.LessThan(minSupportedK8sVersionForPdb)
+		},
+		TestFunc: NodePDBDisablingParallelUpgrade,
+	},
+}
+
+func CreateNodePDBBasic(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+		pxVersion := testutil.GetPortworxVersion(cluster)
+
+		if pxVersion.GreaterThanOrEqual(pxVer3_1_2) {
+			logrus.Infof("Validating Node PDB names and default minAvailable")
+			err := testutil.ValidateNodePDB(cluster, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateUpgradeRetryInterval)
+			require.NoError(t, err)
+		}
+		ci_utils.UninstallAndValidateStorageCluster(cluster, t)
+	}
+
+}
+func CreateNodePDBWithStoragelessNode(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+
+		*cluster.Spec.CloudStorage.MaxStorageNodesPerZone = uint32(3)
+		logrus.Info("Validating PDB with storageless nodes using maxstoragenodesperzone value: ", *cluster.Spec.CloudStorage.MaxStorageNodesPerZone)
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+		ci_utils.UninstallAndValidateStorageCluster(cluster, t)
+
+	}
+}
+
+func MaxNodesAvailableForUpgrade(tc *types.TestCase) func(*testing.T) {
+	return func(t *testing.T) {
+		testSpec := tc.TestSpec(t)
+		cluster, ok := testSpec.(*corev1.StorageCluster)
+		require.True(t, ok)
+		cluster = ci_utils.DeployAndValidateStorageCluster(cluster, ci_utils.PxSpecImages, t)
+		pxVersion := testutil.GetPortworxVersion(cluster)
+
+		if pxVersion.GreaterThanOrEqual(pxVer3_1_2) {
+			err := ci_utils.CordonNodes()
+			require.NoError(t, err)
+
+			logrus.Infof("Validating number of nodes ready for upgrade without minAvailable annotation")
+			err = testutil.ValidateNodesSelectedForUpgrade(cluster, -1, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateUpgradeRetryInterval)
+			require.NoError(t, err)
+
+			k8snodecount, err := ci_utils.GetNonMasterK8sNodeCount()
+			require.NoError(t, err)
+			cluster, err = operator.Instance().GetStorageCluster(cluster.Name, cluster.Namespace)
+			require.NoError(t, err)
+			cluster.Annotations["portworx.io/storage-pdb-min-available"] = fmt.Sprintf("%d", k8snodecount-1)
+			cluster, err = ci_utils.UpdateStorageCluster(cluster)
+			require.NoError(t, err)
+
+			logrus.Infof("Validating number of nodes ready for upgrade with minAvailable annotation %d", k8snodecount-1)
+			err = testutil.ValidateNodesSelectedForUpgrade(cluster, k8snodecount-1, ci_utils.DefaultValidateUpgradeTimeout, ci_utils.DefaultValidateUpgradeRetryInterval)
```

Comment:
In the `ValidateNodesSelectedForUpgrade` function, we are using the `k8snodecount` as a quorum count. Which means we will try to validate upgradable nodes count against `k8snodecount/2 + 1` instead of `totalStorageNodes - k8snodecount`.

Reply:
So when we pass k8snodecount-1 as quorum, we directly use it as the minAvailable value. I realised the variable name is confusing. I will change it to minAvailable. But yeah, I did realise the issue here. I should check if that value is lesser than totalcount-quorum and not just quorum